api/rest/client/lbclient.go

```diff
@@ -0,0 +1,419 @@
+package client
+
+import (
+	"context"
+	"sync/atomic"
+
+	cid "github.com/ipfs/go-cid"
+	shell "github.com/ipfs/go-ipfs-api"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/ipfs-cluster/api"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// loadBalancingClient is a client to interact with IPFS Cluster APIs
+// that balances the load by distributing requests among peers.
+type loadBalancingClient struct {
+	strategy LBStrategy
+	retries  int
+}
+
+// LBStrategy is a strategy to load balance requests among clients.
+type LBStrategy interface {
+	Next(count int) Client
+	SetClients(clients []Client)
+}
+
+// RoundRobin is a load balancing strategy that would use clients in a sequence.
+type RoundRobin struct {
+	clients []Client
+	counter uint32
+}
+
+// Next return the next client to be used.
+func (r *RoundRobin) Next(count int) Client {
+	i := atomic.AddUint32(&r.counter, 1)
+
+	return r.clients[i]
+}
+
+// SetClients sets a list of clients for this strategy.
+func (r *RoundRobin) SetClients(cl []Client) {
+	r.clients = cl
+}
+
+// Failover is a load balancing strategy that would try the local cluster peer first.
+// If the local call fail it would try other client in a round robin fashion.
+type Failover struct {
+	clients []Client
+	counter uint32
+}
+
+// Next returns the next client to be used.
+func (f *Failover) Next(count int) Client {
+	if count == 0 {
+		return f.clients[0]
+	}
+
+	i := atomic.AddUint32(&f.counter, 1)
+
+	if i == 0 {
+		return f.Next(count)
+	}
+
+	return f.clients[i]
+}
+
+// SetClients sets a list of clients for this strategy.
+func (f *Failover) SetClients(cl []Client) {
+	f.clients = cl
+}
+
+// NewLBClient returens a new client that would load balance requests among
+// clients.
+func NewLBClient(strategy LBStrategy, cfgs []*Config, retries int) (Client, error) {
+	var clients []Client
+	for _, cfg := range cfgs {
+		defaultClient, err := NewDefaultClient(cfg)
+		if err != nil {
+			return nil, err
+		}
+		clients = append(clients, defaultClient)
+	}
+	strategy.SetClients(clients)
+	return &loadBalancingClient{strategy: strategy, retries: retries}, nil
+}
+
+// retry tries the request until it is successful or tries `lc.retries` times.
+func (lc *loadBalancingClient) retry(count int, call func(Client) error) error {
+	logger.Debugf("retrying %d times", count+1)
+
+	err := call(lc.strategy.Next(count))
+	count++
+
+	// successful request
+	if err == nil {
+		return nil
+	}
+
+	apiErr, ok := err.(*api.Error)
+	if !ok {
```

Comment:
I think this is wrong. Auto-retries should happen when the error is NOT an `*api.Error` (you don't want to retry on 404s.

Reply:
Looking at client.do,
All errors returned by client methods are of type `api.Error`, even if error is not coming from the api as response (i.e, something to do with the client or how we are making requests). We check the code to differentiate. If code is 0, then it is from the client and not from API.
https://github.com/ipfs/ipfs-cluster/blob/53f1a99feb8f070b87fec3eb0c8eda13f555f211/api/rest/client/request.go#L28

It is a safety check. This error should never occur. It must always cast into api.Error.

Reply:
ok but then:

```
if apiErr.Code != 0 {
		return err
}
```

should come earlier because you don't want to check retry count if you're not planning to retry?

Maybe put a comment to indicate the safeguard too.

Also note that you don't need to check for `ok`, if it does not cast, `apiErr.Code` will stay as 0 and you can retry, so everything can happen with a single check.